src/main/java/com/arpnetworking/metrics/common/sources/KafkaSource.java

```diff
@@ -0,0 +1,287 @@
+import net.sf.oval.constraint.CheckWithCheck;
+import net.sf.oval.constraint.NotNull;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.KafkaException;
+
+import java.time.Duration;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Produce instances of {@code Record} from the values of entries
+ * from a Kafka topic. The key from the entries gets discarded
+ *
+ * @param <T> the type of data created by the source
+ * @param <V> the type of data of value in kafka {@code ConsumerRecords}
+ *
+ * @author Joey Jackson (jjackson at dropbox dot com)
+ */
+public final class KafkaSource<T, V> extends BaseSource {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaSource.class);
+
+    private final Consumer<?, V> _consumer;
+    private final RunnableConsumer _runnableConsumer;
+    private final ExecutorService _consumerExecutor;
+    private final Parser<T, V> _parser;
+    private final Logger _logger;
+    private final Duration _shutdownAwaitTime;
+    private final Duration _backoffTime;
+
+    @Override
+    public void start() {
+        _consumerExecutor.execute(_runnableConsumer);
+    }
+
+    @Override
+    public void stop() {
+        _runnableConsumer.stop();
+
+        _consumerExecutor.shutdown();
+        try {
+            _consumerExecutor.awaitTermination(_shutdownAwaitTime.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (final InterruptedException e) {
+            LOGGER.warn()
+                    .setMessage("Unable to shutdown kafka consumer executor")
+                    .setThrowable(e)
+                    .log();
+        }
+    }
+
+    /**
+     * Generate a Steno log compatible representation.
+     *
+     * @return Steno log compatible representation.
+     */
+    @LogValue
+    public Object toLogValue() {
+        return LogValueMapFactory.builder(this)
+                .put("super", super.toLogValue())
+                .put("consumer", _consumer)
+                .build();
+    }
+
+    @Override
+    public String toString() {
+        return toLogValue().toString();
+    }
+
+    @SuppressWarnings("unused")
+    private KafkaSource(final Builder<T, V> builder) {
+        this(builder, LOGGER);
+    }
+
+    /* package private */ KafkaSource(final Builder<T, V> builder, final Logger logger) {
+        super(builder);
+        _logger = logger;
+        _consumer = builder._consumer;
+        _parser = builder._parser;
+        _runnableConsumer = new RunnableConsumerImpl.Builder<V>()
+                .setConsumer(builder._consumer)
+                .setListener(new LogConsumerListener())
+                .setPollTime(builder._pollTime)
+                .build();
+        _consumerExecutor = Executors.newSingleThreadExecutor(runnable -> new Thread(runnable, "KafkaConsumer"));
```

Comment:
I think we should make this a fixed thread pool executor with a configurable number of threads.